go/p2p/discovery/bootstrap/client.go

```diff
@@ -0,0 +1,162 @@
+	seed            peer.AddrInfo
+	retentionPeriod time.Duration
+
+	mu         sync.Mutex
+	peers      map[string][]peer.AddrInfo
+	timestamps map[string]time.Time
+}
+
+// NewClient creates a new bootstrap protocol client.
+func NewClient(h host.Host, seed peer.AddrInfo, opts ...ClientOption) discovery.Discovery {
+	l := logging.GetLogger("p2p/discovery/bootstrap").With(
+		"seed_id", seed.ID,
+		"seed_addrs", seed.Addrs,
+	)
+
+	do := ClientOptions{}
+	for _, opt := range opts {
+		opt(&do)
+	}
+
+	// Add the seed to the peer store so that libp2p knows how to contact it.
+	h.Peerstore().AddAddrs(seed.ID, seed.Addrs, peerstore.PermanentAddrTTL)
+
+	rc := rpc.NewClient(h, ProtocolID())
+
+	return &client{
+		logger:          l,
+		host:            h,
+		rc:              rc,
+		seed:            seed,
+		retentionPeriod: do.retentionPeriod,
+		peers:           make(map[string][]peer.AddrInfo),
+		timestamps:      make(map[string]time.Time),
+	}
+}
+
+// Advertise implements discovery.Advertiser and discovery.Discovery.
+func (c *client) Advertise(ctx context.Context, ns string, opts ...discovery.Option) (time.Duration, error) {
+	req := AdvertiseRequest{
+		Namespace: ns,
+	}
+	var res AdvertiseResponse
+
+	_, err := c.rc.Call(ctx, c.seed.ID, MethodAdvertise, req, &res)
+	if err != nil {
+		c.logger.Error("failed to advertise",
+			"err", err,
+		)
+
+		return 0, fmt.Errorf("failed to advertise: %w", err)
+	}
+
+	return res.TTL, nil
+}
+
+// FindPeers implements discovery.Discoverer and discovery.Discovery.
+func (c *client) FindPeers(ctx context.Context, ns string, opts ...discovery.Option) (<-chan peer.AddrInfo, error) {
+	// Check how many peers we need to fetch.
+	var options discovery.Options
+	err := options.Apply(opts...)
+	if err != nil {
+		return nil, fmt.Errorf("failed to apply options: %w", err)
+	}
+	limit := options.Limit
+
+	// Lock also RPCs to ensure only one request per retention period.
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if c.timestamps[ns].Add(c.retentionPeriod).Before(time.Now()) {
+		peers, err := c.findPeersLocked(ctx, ns, limit)
+		if err != nil {
+			c.logger.Error("failed to find peers",
+				"err", err,
+			)
+
+			return nil, fmt.Errorf("failed to find peers: %w", err)
+		}
+
+		c.peers[ns] = peers
+		c.timestamps[ns] = time.Now()
+	}
+
+	// Send a random subset to a channel.
+	peers := c.peers[ns]
+	if limit == 0 || limit > len(peers) {
+		limit = len(peers)
+	}
+
+	ch := make(chan peer.AddrInfo, limit)
+	defer close(ch)
+
+	for _, i := range rand.Perm(len(peers))[:limit] {
+		ch <- peers[i]
+	}
+
+	return ch, nil
+}
+
+func (c *client) findPeersLocked(ctx context.Context, ns string, limit int) ([]peer.AddrInfo, error) {
+	req := DiscoverRequest{
+		Namespace: ns,
+		Limit:     limit,
+	}
+	var rsp DiscoverResponse
+
+	_, err := c.rc.Call(ctx, c.seed.ID, MethodDiscover, req, &rsp)
+	if err != nil {
+		return nil, fmt.Errorf("failed to call seed node: %w", err)
+	}
+
+	peers := make([]peer.AddrInfo, len(rsp.Peers))
```

Comment:
Not sure about punishing the node as we will probably have only few seeds. If this was PEX, then I would punish/block the peer. And yes, I should stop processing them although it feels like I'm throwing away the data.

Reply:
This would be a protocol violation, no? If you make a request and say give me at most 10 peers, but the node is giving you 1000 peers, then there is either a bug or someone is not following the protocol which is to respect the limit field. If you want more then you should set a higher limit in which case it is fine.

Reply:
Yes, it's protocol violation. Will just punish it as blocking could left us with no seeds.